config_test.go

```diff
@@ -85,21 +143,230 @@ func TestE2E(t *testing.T) {
+	}
+
+	caFile, err := ioutil.TempFile("", "CA")
+	if err != nil {
+		return "", fmt.Errorf("failed to create temp CA file: %s", err)
+	}
+	defer caFile.Close()
+
+	if err := ioutil.WriteFile(caFile.Name(), caBytes, 0666); err != nil {
+		return "", fmt.Errorf("failed to write CA file: %s", err)
+	}
+
+	return caFile.Name(), nil
+}
+
+func generateKeypairToTempFilesFromCA(ca *certtest.Authority) (string, string, error) {
+	cert, err := ca.BuildSignedCertificate("cert")
+	if err != nil {
+		return "", "", fmt.Errorf("failed to make certificate keypair: %s", err)
+	}
+
+	certBytes, keyBytes, err := cert.CertificatePEMAndPrivateKey()
+	if err != nil {
+		return "", "", fmt.Errorf("failed to get cert and key bytes: %s", err)
+	}
+
+	keyFile, err := ioutil.TempFile("", "key")
+	if err != nil {
+		return "", "", fmt.Errorf("failed to create temp key file: %s", err)
+	}
+	defer keyFile.Close()
+
+	if err := ioutil.WriteFile(keyFile.Name(), keyBytes, 0666); err != nil {
+		return "", "", fmt.Errorf("failed to write key file: %s", err)
+	}
+
+	certFile, err := ioutil.TempFile("", "cert")
+	if err != nil {
+		return "", "", fmt.Errorf("failed to create temp cert file: %s", err)
+	}
+	defer certFile.Close()
+
+	if err := ioutil.WriteFile(certFile.Name(), certBytes, 0666); err != nil {
+		return "", "", fmt.Errorf("failed to write cert file: %s", err)
+	}
+
+	return certFile.Name(), keyFile.Name(), nil
+}
+
+func TestCAInvalidFails(t *testing.T) {
+	t.Parallel()
+
+	invalidCAFile, err := ioutil.TempFile("", "invalid-CA")
+	if err != nil {
+		t.Fatalf("failed to create temp invalid-CA: %v", err)
+	}
+	defer invalidCAFile.Close()
+
+	if err := ioutil.WriteFile(invalidCAFile.Name(), []byte("invalid"), 0666); err != nil {
+		t.Fatalf("failed to write invalid-CA file: %v", err)
+	}
+	defer os.Remove(invalidCAFile.Name())
+
+	_, clientCAErr := tlsconfig.Build().Client(tlsconfig.WithAuthorityFromFile(invalidCAFile.Name()))
+	_, serverCAErr := tlsconfig.Build().Server(tlsconfig.WithClientAuthenticationFromFile(invalidCAFile.Name()))
+
+	buildResults := []buildResult{
+		{name: "CA cert file invalid (client)", err: clientCAErr},
+		{name: "CA cert file invalid (server)", err: serverCAErr},
+	}
+
+	for _, br := range buildResults {
+		br := br // capture variable
+		t.Run(br.name, func(t *testing.T) {
+			t.Parallel()
+
+			if br.err == nil {
+				t.Fatal("building config should have errored")
+			}
+			if !strings.HasPrefix(br.err.Error(), "Unable to load caCert") {
+				t.Fatalf("unexpected error prefix returned; have: %v, want: 'Unable to load caCert", br.err)
+			}
+		})
+	}
+}
+
 func TestInternalDefaults(t *testing.T) {
 	log.SetOutput(ioutil.Discard)
 	t.Parallel()
 
+	clientConfig, err := tlsconfig.Build(tlsconfig.WithInternalServiceDefaults()).Client()
+	if err != nil {
+		t.Fatalf("failed to build client config: %v", err)
+	}
+	serverConfig, err := tlsconfig.Build(tlsconfig.WithInternalServiceDefaults()).Server()
```

Comment:
Meta comment about all of these extractions: it's a shame that the test setup is no longer encapsulated in the test case struct. Adding a new test case requires changes elsewhere in the function rather than just adding a new test case to the table. Maybe we could add new fields to the test case for TLSOptions, ServerOptions, and ClientOptions and then build the configuration as part of the test itself rather than in setup.

Reply:
Are you suggesting we structure the tests like ginkgo table tests?

Reply:
Like the standard Go pattern they're based on: https://github.com/golang/go/wiki/TableDrivenTests

Something like:

```
[]struct {
    name   string
    tlsOpts []TLSOption
    serverOpts []ServerOption
    clientOpts []ClientOption
}{...}
```

may work? It doesn't let you encapsulate the file system setup though.

Reply:
How do you decide when to call the `Client()` or `Server()` final build methods since they are methods not functions?

You could add an option to the test case that tells you if the test case is for the Server or Client but I don't love how that ends up looking in the actual test block, now there's conditional logic in the stuff that is supposed to be common. I think this would *look* better in Ginkgo contexts or if we didn't have to return errors from the `Server()` and `Client()` methods

Reply:
The presence of either clientOpts or serverOpts is mutually exclusive so it should be possible to know which constructor to call based on their presence.

Reply:
@xoebus how would you feel about addressing this as a separate issue (tracker/github/etc). This PR is semi blocking some other work related to our CI pipeline.